sphinx/testing/internal/markers.py

```diff
@@ -0,0 +1,306 @@
+    reflects the signature of the :func:`pytest.mark.sphinx` marker, but
+    not of the :class:`~sphinx.testing.util.SphinxTestApp` constructor.
+    """
+
+    buildername: str
+    confoverrides: dict[str, Any]
+    # using freshenv=True will be treated as equivalent to use isolate=True
+    # but in the future, we might want to deprecate this marker keyword in
+    # favor of "isolate" (that way, we don't need to maintain it)
+    freshenv: bool
+    warningiserror: bool
+    tags: list[str]
+    verbosity: int
+    parallel: int
+    keep_going: bool
+    docutils_conf: str
+
+    # added or updated fields
+    testroot: str | None
+    isolate: NormalizableIsolation
+
+
+class SphinxInitKwargs(TypedDict, total=False):
+    """The type of the keyword arguments after processing.
+
+    Such objects are constructed from :class:`SphinxMarkEnviron` objects.
+    """
+
+    # :class:`sphinx.application.Sphinx` positional arguments as keywords
+    buildername: Required[str]
+    """The deduced builder name."""
+    # :class:`sphinx.application.Sphinx` required arguments
+    srcdir: Required[Path]
+    """Absolute path to the test sources directory.
+
+    The uniqueness of this path depends on the isolation policy,
+    the location of the test and the application's configuration.
+    """
+    # :class:`sphinx.application.Sphinx` optional arguments
+    confoverrides: dict[str, Any] | None
+    status: StringIO | None
+    warning: StringIO | None
+    freshenv: bool
+    warningiserror: bool
+    tags: list[str] | None
+    verbosity: int
+    parallel: int
+    keep_going: bool
+    # :class:`sphinx.testing.util.SphinxTestApp` optional arguments
+    docutils_conf: str | None
+    builddir: Path | None
+    # :class:`sphinx.testing.util.SphinxTestApp` extras arguments
+    isolate: Required[Isolation]
+    """The deduced isolation policy."""
+    testroot: Required[str | None]
+    """The deduced testroot ID (possibly None if the default ID is not set)."""
+    testroot_path: Required[str | None]
+    """The absolute path to the testroot directory, if any."""
+    shared_result: Required[str | None]
+    """The optional shared result ID."""
+
+
+class AppParams(NamedTuple):
+    """The processed arguments of :func:`pytest.mark.sphinx`.
+
+    The *args* and *kwargs* values can be directly used as inputs
+    to the :class:`~sphinx.testing.util.SphinxTestApp` constructor.
+    """
+
+    args: list[Any]
+    """The constructor positional arguments, except ``buildername``."""
+    kwargs: SphinxInitKwargs
+    """The constructor keyword arguments, including ``buildername``."""
+
+
+class TestParams(TypedDict):
+    """A view on the arguments of :func:`pytest.mark.test_params`."""
+
+    shared_result: str | None
+
+
+def _get_sphinx_environ(node: PytestNode, default_builder: str) -> SphinxMarkEnviron:
+    args, kwargs = get_mark_parameters(node, 'sphinx')
+
+    if len(args) > 1:
```

Comment:
For the marker, I think we should stick to positional builder name + keyword arguments for the rest.